homeassistant/components/sma_manager/SMA.py

```diff
@@ -0,0 +1,180 @@
+"""SMA Manager API"""
+
+#  Copyright (c) 2023.
+#  All rights reserved to the creator of the following script/program/app, please do not
+#  use or distribute without prior authorization from the creator.
+#  Creator: Antonio Manuel Nunes Goncalves
+#  Email: amng835@gmail.com
+#  LinkedIn: https://www.linkedin.com/in/antonio-manuel-goncalves-983926142/
+#  Github: https://github.com/DEADSEC-SECURITY
+
+# Built-In Imports
+import socket
+import struct
+from datetime import timedelta
+from typing import Any
+
+SMA_BUFFER_SIZE = 10240
+
+
+def create_multicast_socket(ip: str, port: int, timeout: int = 5):
+    """
+    Creates and returns a socket
+    """
+
+    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
+    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+    sock.bind(("", port))
+    mreq = struct.pack("4sl", socket.inet_aton(ip), socket.INADDR_ANY)
+    sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)
+    sock.settimeout(timeout)
+    return sock
+
+
+def _bytes_to_power(bytes_):
+    """
+    Converts bytes data into a integer representing power in watts
+
+    @param bytes_:
+    @return:
+    """
+
+    return int.from_bytes(bytes_, byteorder="big") / 10
+
+
+class SMA:
+    # Passed variables
+    _ip: str
+    _port: int
+    _refresh_time: timedelta
+
+    _data: Any
+    # Metadata
+    _available: bool
+    _sma: str
+    _uid: str
+    _serial_number: str
+    # Sensor data
+    _grid_consumption: float
+    _grid_feed: float
+    _phase_one_consumption: float
+    _phase_one_feed: float
+    _phase_two_consumption: float
+    _phase_two_feed: float
+    _phase_three_consumption: float
+    _phase_three_feed: float
+
+    def __init__(self, name: str, ip: str, port: int, refresh_time: int = 1):
+        self._name = name
+        self._ip = ip
+        self._port = port
+        self._refresh_time = timedelta(seconds=refresh_time)
+
+        self._load_metadata()
+
+    def _load_metadata(self):
+        self._load_data()
+        self._sma = self._data[0:3].decode("UTF-8")
+        self._uid = self._data[4:7].hex()
+        self._serial_number = self._data[20:24].hex()
```

Comment:
might worth adding documentation on the structure of the packet (if available) as it looks like a lot of magic number.

Reply:
Its part of the device docs from the company. Ill make some documentation on it.